File: 2015/go/5/main.go
package main

import (
	"fmt"
	"os"
)

func main() {
	args := os.Args[1:]

	if len(args) < 1 {
		fmt.Println("Usage: go run main.go <input.txt>")
		return
	}

	file, err := os.Open(args[0])
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}

	defer file.Close()

	part1_count := 0
	part2_count := 0

	// Read the file
	for {
		var line string
		n, err := fmt.Fscanln(file, &line)
		if n == 0 || err != nil {
			break
		}

		if check_good_string_part1(line) {
			part1_count++
		}

		if check_good_string_part2(line) {
			part2_count++
		}
	}

	fmt.Println("Part 1:", part1_count)
	fmt.Println("Part 2:", part2_count)

}

func check_good_string_part1(s string) bool {
	vowels := 0
	double_letter := false
	bad_string := false

	for i := 0; i < len(s); i++ {
		if s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' || s[i] == 'u' {
			vowels++
		}

		if i < len(s)-1 && s[i] == s[i+1] {
			double_letter = true
		}

		if i < len(s)-1 && (s[i] == 'a' && s[i+1] == 'b' || s[i] == 'c' && s[i+1] == 'd' || s[i] == 'p' && s[i+1] == 'q' || s[i] == 'x' && s[i+1] == 'y') {
			bad_string = true
		}
	}

	if vowels >= 3 && double_letter && !bad_string {
		return true
	}

	return false
}

func check_good_string_part2(s string) bool {
	pair := false
	repeat := false

	for i := 0; i < len(s)-1; i++ {

		if i < len(s)-2 && s[i] == s[i+2] {
			repeat = true
		}

		for j := i + 2; j < len(s)-1; j++ {
			if s[i] == s[j] && s[i+1] == s[j+1] {
				pair = true
			}
		}
	}

	return pair && repeat
}

File: 2015/php/5/main.php
<?php

function check_good_string_part1($string) {
    $vowels = 0;
    $double_letter = false;
    $bad_string = false;

    for($i = 0; $i < strlen($string); $i++) {
        if ($string[$i] == 'a' || $string[$i] == 'e' || $string[$i] == 'i' || $string[$i] == 'o' || $string[$i] == 'u') {
            $vowels++;
        }

        if ($i > 0) {
            if ($string[$i] == $string[$i - 1]) {
                $double_letter = true;
            }
            if ($string[$i] == 'b' && $string[$i - 1] == 'a') {
                $bad_string = true;
            }
            if ($string[$i] == 'd' && $string[$i - 1] == 'c') {
                $bad_string = true;
            }
            if ($string[$i] == 'q' && $string[$i - 1] == 'p') {
                $bad_string = true;
            }
            if ($string[$i] == 'y' && $string[$i - 1] == 'x') {
                $bad_string = true;
            }
        }
    }
    
    return $vowels >= 3 && $double_letter && !$bad_string;
}

function check_good_string_part2($string){
    $pair = false;
    $repeat = false;

    for($i = 0; $i < strlen($string); $i++) {
        if ($i > 0) {
            $pair = $pair || strpos($string, substr($string, $i - 1, 2), $i + 1) !== false;
            $repeat = $repeat || $string[$i] == $string[$i - 2];
        }
    }

    return $pair && $repeat;

}

function main(){
    $args = $_SERVER['argv'];

    if (count($args) < 2) {
        echo "Usage: php main.php <input.txt>\n";
        exit(1);
    }

    $filename = $args[1];

    $file = fopen($filename, "r") or die("Unable to open file!");
    
    if ($file == false) {
        echo "Error: Unable to open file.\n";
        exit(1);
    }

    $all_lines = array();
    while (!feof($file)) {
        $line = fgets($file);
        $all_lines[] = $line;
    }

    fclose($file);

    $part1_count = 0;
    $part2_count = 0;

    for ($i = 0; $i < count($all_lines); $i++) {
        $line = $all_lines[$i];
        
        if (check_good_string_part1($line)) {
            $part1_count++;
        }

        if (check_good_string_part2($line)) {
            $part2_count++;
        }
    }

    echo "Part 1: $part1_count\n";
    echo "Part 2: $part2_count\n";

}

main();